### PR TITLE
Chris - convert analytics to TS

### PR DESCRIPTION
**Pull request by chreamy**

### CODE REVIEW

File: packages/analytics/package.json [hunk 1]
```diff
@@ -6,9 +6,10 @@
   "type": "module",
   "license": "GPL",
   "scripts": {
-    "start": "node src/index.js",
+    "start": "node --loader ts-node/esm src/index.ts",
```
Comment (lancatlin) [inline packages/analytics/package.json:9]: "start" is for production execution, let it execute the compiled js.

File: packages/analytics/src/controllers/commands.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
 /* eslint-disable object-shorthand */
-import Pagination from '../models/pagination.js'
+import Pagination from '../models/pagination.ts'
```
Comment (lancatlin) [inline packages/analytics/src/controllers/commands.ts:2]: use ".js" in import

File: packages/analytics/src/controllers/commands.ts [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
 import { errorCommands } from '../models/commands.js'
 
 // eslint-disable-next-line import/prefer-default-export
-export async function commandListController(req, res) {
-  const { nextKey, order } = req.query
+export async function commandListController(req: any, res: any): Promise<void> {
```
Comment (lancatlin) [inline packages/analytics/src/controllers/commands.ts:6]:
```suggestion
import { Request, Response } from 'express'
export async function commandListController(req: Request, res: Response): Promise<void> {
```

File: packages/analytics/src/models/sessions.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import { Types } from 'mongoose'
+
+function formatTime(time: number) {
+  const seconds = Math.floor(time / 1000)
+  const minutes = Math.floor(seconds / 60)
+  const hours = Math.floor(minutes / 60)
+  return `${hours}h ${minutes % 60}m ${seconds % 60}s`
+}
+
+export async function sessionList(pagination: Pagination): Promise<any[]> {
```
Comment (lancatlin) [inline packages/analytics/src/models/sessions.ts:12]: Prevent using any
```suggestion
import { ISession } from '@linux-odyssey/models'

export async function sessionList(pagination: Pagination): Promise<ISession[]> {
```
ref: [Mongoose TypeScript support](https://mongoosejs.com/docs/typescript.html)

Comment (lancatlin) [reply packages/analytics/src/models/sessions.ts:12]: Oh, I'm wrong. It's a custom type actually. Create a interface or type for that.

File: server/src/containers/docker.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import Docker from 'dockerode'
 import config, { getQuestImage } from '../config.js'
 import logger from '../utils/logger.js'
 
-const engine = new Docker()
+const engine = new Docker({ socketPath: '//./pipe/docker_engine' })
```
Comment (lancatlin) [inline server/src/containers/docker.ts:5]: Why is this?

File: server/src/containers/docker.ts [hunk 6]
```diff
@@ -62,7 +62,7 @@ export async function attachContainer(
     Tty: true,
     Env: [
       `TOKEN=${token}`,
-      `API_ENDPOINT=${config.backendUrl}`,
+      `API_ENDPOINT=http://backend:3000`,
```
Comment (lancatlin) [inline server/src/containers/docker.ts:65]: change environment variables in your personal `.env` instead of hard coding. You may need to import `dotenv` for load `.env` files

**Review by lancatlin — CHANGES_REQUESTED** [review]
Prevent using `any`

File: server/src/containers/docker.ts (hunk 5 above)
Comment (chreamy) [reply server/src/containers/docker.ts:5]: locally testing, im reverting it back

File: packages/analytics/src/controllers/commands.ts (hunk 3 above)
Comment (chreamy) [reply packages/analytics/src/controllers/commands.ts:6]: Just did

File: packages/analytics/src/models/sessions.ts (hunk 4 above)
Comment (chreamy) [reply packages/analytics/src/models/sessions.ts]: I created the interfaces under analytics/src/interface.ts

File: packages/analytics/src/controllers/commands.ts (hunk 2 above)
Comment (lancatlin) [reply packages/analytics/src/controllers/commands.ts:2]: What's going on here? Why you redo those changes?

File: docker-compose.dev.yml [hunk 7]
```diff
@@ -78,7 +78,7 @@ services:
     networks:
       - internal
     ports:
-      - '27017:27017'
+      - '27018:27017'
```
Comment (lancatlin) [inline docker-compose.dev.yml:81]: Revert this change

File: package-lock.json
Comment (lancatlin) [inline package-lock.json:1]: Remove this file. We use Yarn

File: packages/analytics/nodemon.json [hunk 8]
```diff
@@ -1,3 +1,6 @@
 {
-  "ext": "js,handlebars"
+  "watch": ["src", "views"],
+  "ext": "ts,handlebars,js",
+  "ignore": ["src/**/*.spec.ts"],
+  "exec": "node --loader ts-node/esm src/index.js"
```
Comment (lancatlin) [inline packages/analytics/nodemon.json:5]:
```suggestion
  "exec": "node --loader ts-node/esm src/index.ts"
```

File: packages/analytics/package.json [hunk 9]
```diff
@@ -6,19 +6,27 @@
...
     "build": "tsc",
-    "dev": "nodemon src/index.js"
+    "watch": "tsc -w",
+    "dev": "nodemon --loader ts-node/esm src/index.ts --trace-warnings"
```
Comment (lancatlin) [inline packages/analytics/package.json:12]:
```suggestion
    "dev": "nodemon"
```
We defined nodemon config so stick with that.

File: packages/analytics/package.json (hunk 1 above)
Comment (lancatlin) [reply packages/analytics/package.json:9]: Why revert it?

File: packages/analytics/package.json [hunk 10]
```diff
@@ -6,19 +6,27 @@
   "type": "module",
   "license": "GPL",
   "scripts": {
-    "start": "node src/index.js",
+    "start": "node --loader ts-node/esm src/index.ts",
```
Comment (lancatlin) [inline packages/analytics/package.json:9]:
```suggestion
    "start": "node dist/index.js",
```

File: packages/analytics/src/interface.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  user: User
+  quest: string
+  status: string
+  createdAt: Date
+  finishedAt: Date
+}
+export interface CommandObject {
+  command: string
+  pwd: string
+  stage: string | null | undefined
```
Comment (lancatlin) [inline packages/analytics/src/interface.ts:17]:
```suggestion
  stage?: string | null
```
Use `?:` for optional field

**Review by lancatlin — CHANGES_REQUESTED** [review]

**Review by chreamy — COMMENTED** [review]
Should be good now
